### PR TITLE
A grid company must be able to link a child metering point to a parent metering point

### PR DESCRIPTION
**Pull request by Korgath**
## References

https://app.zenhub.com/workspaces/batman-60a6105157304f00119be86e/issues/energinet-datahub/geh-metering-point/301

### CODE REVIEW

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/Errors/Converters/Create/Special/ParentMustBeMarketMeteringPointConditionalRuleErrorConverter.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Energinet.DataHub.MeteringPoints.Infrastructure.EDI.Errors.Converters.Create.Special
+{
+    public class ParentMustBeMarketMeteringPointConditionalRuleErrorConverter : ErrorConverter<ParentMustBeMarketMeteringPointConditionalRuleError>
+    {
+        protected override ErrorMessage Convert(ParentMustBeMarketMeteringPointConditionalRuleError validationError)
+        {
+            if (validationError == null) throw new ArgumentNullException(nameof(validationError));
+
+            return new ErrorMessage("D18", "Only a metering point in group 3 and 4 can be a child of 17 and 18 metering point.");
```
Comment (erduna) [inline source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/Errors/Converters/Create/Special/ParentMustBeMarketMeteringPointConditionalRuleErrorConverter.cs:26]: Consumption and production, not 17/18

File: source/Energinet.DataHub.MeteringPoints.IntegrationTests/CreateMeteringPoints/SpecialMeteringPointTests.cs [hunk 2]
```diff
@@ -31,14 +33,90 @@ public SpecialMeteringPointTests(DatabaseFixture databaseFixture)
...
+
+            var specialCommand = CreateCommand()
+                with { ParentRelatedMeteringPoint = SampleData.SecondGsrnNumber };
+            await SendCommandAsync(specialCommand).ConfigureAwait(false);
+
+            AssertValidationError("D18", DocumentType.CreateMeteringPointRejected);
+        }
+
+        [Fact]
+        public async Task Should_accept_if_parent_metering_point_is_not_a_market_metering_point_when_type_is_ExchangeReactiveEnergy()
```
Comment (erduna) [inline source/Energinet.DataHub.MeteringPoints.IntegrationTests/CreateMeteringPoints/SpecialMeteringPointTests.cs:50]: Specify If it is exchange. Only that one can be parent of reactive.

File: source/Energinet.DataHub.MeteringPoints.IntegrationTests/CreateMeteringPoints/SpecialMeteringPointTests.cs [hunk 3]
```diff
@@ -31,14 +33,90 @@ public SpecialMeteringPointTests(DatabaseFixture databaseFixture)
...
+                    ParentRelatedMeteringPoint = SampleData.SecondGsrnNumber,
+                    MeteringPointType = nameof(MeteringPointType.ExchangeReactiveEnergy),
                 };
+            await SendCommandAsync(specialCommand).ConfigureAwait(false);
+
+            AssertOutboxMessage<MessageHubEnvelope>(envelope => envelope.MessageType == DocumentType.CreateMeteringPointAccepted, 2);
+        }
+
+        [Fact]
+        public async Task Should_not_generate_integration_event_message_if_parentId_is_null()
```
Comment (erduna) [inline source/Energinet.DataHub.MeteringPoints.IntegrationTests/CreateMeteringPoints/SpecialMeteringPointTests.cs:68]: Are we sure it should not? An integration event would then be generated for a parent coupling through update MP, but that event does not contain enough information. I’d argue we should trigger the event but with parent of course as null.

**Review by erduna — CHANGES_REQUESTED** [review]
I’d argue integration event needs to be triggered even without a parent attached, as it contains information no other events from eg. Update MP does.

File: source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/Special/Rules/GridAreaMustBeTheSameForParentRule.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+namespace Energinet.DataHub.MeteringPoints.Domain.MeteringPoints.Special.Rules
+{
+    public class GridAreaMustBeTheSameForParentRule : IBusinessRule
+    {
+        public GridAreaMustBeTheSameForParentRule(bool parentIsInSameGridArea)
+        {
+            IsBroken = !parentIsInSameGridArea;
+            ValidationError = new GridAreaMustBeTheSameForParentRuleError();
+        }
+
```
Comment (kft) [inline source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/Special/Rules/GridAreaMustBeTheSameForParentRule.cs:26]: Is this rule necessary ? There is no logic. Could this be handled at client ?

**Review by kft — DISMISSED** [review]

**Pull request closed without merging**